### ISSUE TITLE
/usr/bin/python3 required even with a hermetic python toolchain registered.

### ISSUE THREAD

**Issue by jacky8hyf** [bazelbuild/bazel#19355]
### Description of the bug:

This is a bit similar to #8685, but the request is slightly different. I registered a Python toolchain with prebuilts in the source tree, but the bootstrap script for a `py_binary` still expects /usr/bin/python3 to exist. 

### Which category does this issue belong to?

Python Rules

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

BUILD.bazel:
```
load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")

py_runtime(
    name = "python3",
    files = ["py3-cmd"],
    interpreter = "py3-cmd",
    python_version = "PY3",
)

py_runtime_pair(
    name = "py_runtime_pair",
    py3_runtime = ":python3",
)

toolchain(
    name = "py_toolchain",
    toolchain = ":py_runtime_pair",
    toolchain_type = "@bazel_tools//tools/python:toolchain_type",
)

py_binary(
    name = "hello",
    srcs = ["hello.py"],
    main = "hello.py",
)
```

hello.py:
```
print("hello!")
```

WORKSPACE:
```
register_toolchains("//:py_toolchain")
```

py3-cmd (This is our fake Python runtime for illustrative purposes. In practice, this should be replaced by a real Python interpreter.)
```
#!/bin/bash

echo "PYTHON" "$@"
```

With this setup, run:

```
$ chmod +x py3-cmd

# This will run successfully
$ bazel run :hello
PYTHON <...>/47bd642ed3e6d055ba60063dc591a588/execroot/__main__/bazel-out/k8-fastbuild/bin/hello.runfiles/__main__/hello.py

# Delete /usr/bin/python3
sudo mv /usr/bin/python3 /usr/bin/python3.bak

# This now fails
$ bazel run :hello
/usr/bin/env: ‘python3’: No such file or directory
```

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 6.3.2

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

No

### Have you found anything relevant by searching the web?

#8685

Note that in this case, `#!/usr/bin/env python3` is the default shebang of the py_runtime:

https://bazel.build/reference/be/python#py_runtime.stub_shebang

A possible workaround would be to set `py_runtime.stub_shebang` to `#!/<path to workspace>/py3-cmd`. However, this solution:

- Requires inferring the workspace path in the analysis phase
- Writes the absolute path to workspace

Alternatively I can also set `#!py3-cmd`.

But in both cases, I think it is counter-intuitive that, even after I register a Python toolchain to be a hermetic one that I provide via prebuilts, I am still depending on the host machine to install a Python runtime.

**Comment by jacky8hyf** [comment]
This is (sort of) blocking partner's effort to run the build in an nsjail, where they intentionally dropped Python from /usr/bin.

See internal bug 297913707.

**Comment by lberki** [comment]
Channeling @rickeylev , it looks like the Python toolchain you are using is `//prebuilts/build-tools:python3`, which supports the [bootstrap_template](https://bazel.build/reference/be/python#py_runtime.bootstrap_template) attribute that you can use to replace the bootstrap script with one that doesn't depend on python3 or the `stub_shebang` attribute that allows you to replace just the shebang line. Does that maybe help?

**Comment by rickeylev** [comment]
Yes, you need some sort of shebang. This is intrinsic to any program that isn't a native executable (python, java, perl, shell, etc, most any language with a separate "interpreter").

What bootstrap_template lets you do, that stub_shebang doesn't, is replace more than just the shebang. For example, a template like this removes the reliance on having a host Python available:

```
#!/bin/sh
runfiles_dir="${0}.runfiles"
python="$runfiles_dir/%python_binary%"
exec $python %main% $@
```

That would be the basic core of removing the host-python reliance. The `%name%` stuff are the strings replaced by py_binary. I'm not sure how robust `${0}.runfiles` will be in practice. A full implementation would still need to do most of the same bootstrap logic that you see in the existing python_bootstrap_template.txt. You could re-implement that in shell 😬, or do something like:

```
exec "$python" - "$@" <<'EOF'
<copy/paste the existing bootstrap_template.txt, mostly>
'EOF'
```

Internally, we do something similar to that to make some of the Bazel integration tests happy with our custom Python runtime.

Hm. I wonder if the template could be actual native code. Like you pre-build a binary, and embed `%main%\0` in it (right-padding a bunch of null bytes), and Bazel just replaces those placeholders like normal. I'm not sure if non-textual data will survive `ctx.actions.expand_template()`, though. Could be brittle, or amazing, or both.

**Comment by lberki** [comment]
@rickeylev I could pick the "native code template" idea apart (for example, the substitution would change the length of the blob, which probably breaks the binary in it horribly), but I have to admit that it's a delightfully creative idea :)

**Comment by rickeylev** [comment]
The python rules are now hosted in https://github.com/bazel-contrib/rules_python

If this is still an issue, please file an issue there.
